src/metabase/driver/sql_jdbc/actions.clj

```diff
@@ -203,17 +203,72 @@
 (defn- prepare-query [hsql-query driver action]
   (prepare-query* driver action hsql-query))
 
-(defn- row-delete!* [action _context {db-id :database, :as query}]
-  (let [database             (actions/cached-database db-id)
-        ;; Don't trust the driver we dispatched on - it's deprecated anyhow.
+(defmulti do-nested-transaction
+  "Execute `thunk` inside a nested transaction inside `connection`, which is currently in a transaction. If `thunk`
+  throws an Exception, the nested transaction should be rolled back, but the parent transaction should be able to
+  proceed.
+
+  Why do we need this?
+
+  With things like bulk insert, we want to collect all the errors for all the rows in one go. Say you have 4 rows, 1 2
+  3 and 4. If 1 errors then depending on the DBMS, the transaction enters an error state that disallows doing anything
+  else. 2, 3, and 4 will error with a \"transaction has been aborted\" error that you can't clear (AFAIK). This
+  affects Postgres but not H2. Not sure about other DBs yet.
+
+  Without using nested transactions, if you have errors in rows 2 and 4 you'd only see the error in row 2 since 3 and
+  4 would fail with \"transaction has been aborted\" or whatever.
+
+  So the point of using nested transactions is that if 2 is done inside a nested transaction we can rollback the
+  nested transaction which allows the top-level transaction to proceed even tho part of it errored."
+  {:changelog-test/ignore true :arglists '([driver ^java.sql.Connection connection thunk]), :added "0.44.0"}
+  driver/dispatch-on-initialized-driver
+  :hierarchy #'driver/hierarchy)
+
+(defn- run-bulk-transaction!
```

Comment:
This is a simplified version of `perform-bulk-action-with-repeated-single-row-actions!`.

Yes, it's a bit weird that both flavors now support batches. The salient difference is being MBQL based.